_posts/2021-05-30-cloudnative-bot.md

```diff
@@ -0,0 +1,158 @@
+
+The behind the technology that the bot uses will be explained in this blog post.
+
+## Pre-requisites
+
+* [Docker or Docker Desktop](https://www.docker.com/products/docker-desktop)
+* OpenFaaS
+* faas-cli
+* [connector-sdk](https://github.com/openfaas/connector-sdk)
+* [ko](https://github.com/google/ko)
+
+### Using Connector-sdk to listen and dispatch the Twitter Stream
+
+The connector-sdk is very handy and can be used to listen to an event and act if something happens by triggering a specific function.
+
+The good part is you can set up that with a few lines of code. In the case of Twitter, we need to configure to listen to the stream. 
+
+To configure the Twitter stream, we can do something like:
+
+```golang
+api = anaconda.NewTwitterApiWithCredentials(Config.TwitterAccessToken, Config.TwitterAccessSecret, Config.TwitterConsumerKey, Config.TwitterConsumerSecretKey)
+  if _, err := api.VerifyCredentials(); err != nil {
+    log.Fatalf("Bad Authorization Tokens. Please refer to https://apps.twitter.com/ for your Access Tokens: %s", err)
+  }
+
+  streamValues := url.Values{}
+  streamValues.Set("track", "/cloudnativetv,/k8s-patch-schedule,/kubecon-random-video")
+  streamValues.Set("stall_warnings", "true")
+  log.Println("Starting CloudNative Stream...")
+  s := api.PublicStreamFilter(streamValues)
+```
+
+After that, we need to configure the connector-sdk
+
+```golang
+creds := &auth.BasicAuthCredentials{
+    User:     Config.OpenFaasUsername,
+    Password: Config.OpenFaasPassword,
+  }
+
+  config := &types.ControllerConfig{
+    RebuildInterval:         time.Millisecond * 1000,
+    GatewayURL:              Config.OpenFaasGateway,
+    PrintResponse:           true,
+    PrintResponseBody:       true,
+    AsyncFunctionInvocation: false,
+  }
+
+  controller := types.NewController(creds, config)
+
+  receiver := ResponseReceiver{}
+  controller.Subscribe(&receiver)
+
+  controller.BeginMapBuilder()
+```
+
+Then we can create a loop to listen to the events and dispatch when needed.
+
+```golang
+go func() {
+    for t := range s.C {
+      switch v := t.(type) {
+      case anaconda.Tweet:
+        data := []byte(fmt.Sprintf(`{"text": %q,"id": %d, "id_str": %q, "user_screen_name": %q, "api_key": %q}`, v.Text, v.Id, v.IdStr, v.User.ScreenName, Config.TokenAPIKey))
+
+        topic := "cloudnative.twitter.stream"
+        log.Printf("Got one message - https://twitter.com/%s/status/%d - Invoking on topic %s\n", v.User.ScreenName, v.Id, topic)
+        controller.Invoke(topic, &data)
+      default:
+        log.Printf("Got something else %v", v)
+      }
+    }
+  }()
+```
+
+The complete source code can see here: https://github.com/cpanato/cloudnative-bot/blob/main/cmd/twitter-stream/main.go.
+
+To deploy this service, we are using `ko` with is an excellent tool, and we don't need to define a docker file; you can build and push the container right away and have your Kubernetes manifests configured with that image.
```

Comment:
Explain how this is different to the usual workflow. Why is it better or faster or more interesting?

Compare and contrast both approaches (briefly)